plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java

```diff
@@ -0,0 +1,570 @@
+    }
+
+    @Override
+    public ListenableFuture<Void> deleteRecursively(URI dir)
+    {
+        return asVoid(Futures.transformAsync(
+                toListenableFuture(listObjectsRecursively(dir).byPage().collectList().toFuture()),
+                pagedResponseList -> deleteObjects(getContainerName(dir), pagedResponseList),
+                directExecutor()));
+    }
+
+    @Override
+    public List<FileStatus> listFiles(URI dir)
+            throws IOException
+    {
+        ImmutableList.Builder<FileStatus> builder = ImmutableList.builder();
+        try {
+            for (BlobItem blobItem : listObjects(dir)) {
+                if (blobItem.isPrefix() != Boolean.TRUE) {
+                    builder.add(new FileStatus(
+                            new URI(dir.getScheme(), dir.getUserInfo(), dir.getHost(), -1, PATH_SEPARATOR + blobItem.getName(), null, dir.getFragment()).toString(),
+                            blobItem.getProperties().getContentLength()));
+                }
+            }
+        }
+        catch (URISyntaxException e) {
+            throw new IllegalArgumentException(e);
+        }
+        return builder.build();
+    }
+
+    @Override
+    public List<URI> listDirectories(URI dir)
+            throws IOException
+    {
+        ImmutableList.Builder<URI> builder = ImmutableList.builder();
+        try {
+            for (BlobItem blobItem : listObjects(dir)) {
+                if (blobItem.isPrefix() == Boolean.TRUE) {
+                    builder.add(new URI(dir.getScheme(), dir.getUserInfo(), dir.getHost(), -1, PATH_SEPARATOR + blobItem.getName(), null, dir.getFragment()));
+                }
+            }
+        }
+        catch (URISyntaxException e) {
+            throw new IllegalArgumentException(e);
+        }
+        return builder.build();
+    }
+
+    @Override
+    public int getWriteBufferSize()
+    {
+        return blockSize;
+    }
+
+    @PreDestroy
+    @Override
+    public void close()
+            throws IOException
+    {
+    }
+
+    private PagedIterable<BlobItem> listObjects(URI dir)
+    {
+        checkArgument(isDirectory(dir), "listObjects called on file uri %s", dir);
+
+        String containerName = getContainerName(dir);
+        String directoryPath = getPath(dir);
+        if (!directoryPath.isEmpty()) {
+            directoryPath += PATH_SEPARATOR;
+        }
+
+        return blobServiceClient.getBlobContainerClient(containerName).listBlobsByHierarchy(PATH_SEPARATOR, (new ListBlobsOptions()).setPrefix(directoryPath), null);
+    }
+
+    private PagedFlux<BlobItem> listObjectsRecursively(URI dir)
+    {
+        checkArgument(isDirectory(dir), "listObjectsRecursively called on file uri %s", dir);
+
+        String containerName = getContainerName(dir);
+        String directoryPath = getPath(dir);
+
+        return blobServiceAsyncClient.getBlobContainerAsyncClient(containerName).listBlobsByHierarchy(null, (new ListBlobsOptions()).setPrefix(directoryPath));
```

Comment:
How many objects does it return in a single batch by default?

Reply:
Default is 5000.